src/components/UncontrolledTable.tsx

```diff
@@ -0,0 +1,269 @@
+import React, { FC, useCallback, useEffect, useState } from 'react';
+import type { Key } from 'react';
+import orderBy from 'lodash.orderby';
+import some from 'lodash.some';
+import isEqual from 'lodash.isequal';
+import noop from 'lodash.noop';
+import Paginator from './Paginator';
+import SortableTable from './SortableTable/SortableTable';
+import type {
+  SortableColumn,
+  SortableTableProps,
+  SortableTableRow,
+} from './SortableTable/SortableTable';
+
+interface SortProps {
+  column?: Key;
+  ascending: boolean;
+}
+export interface UncontrolledTableProps<T>
+  extends Omit<SortableTableProps<T>, 'rowSelected'> {
+  expandable?: boolean;
+  expanded?: T | T[];
+  onPageChange?: (page: number) => void;
+  onSort?: (sort?: SortProps) => void;
+  onVisibleRowsChange?: (rows: SortableTableRow<T>[]) => void;
+  page?: number;
+  pageSize?: number;
+  paginated?: boolean;
+  selectable?: boolean;
+  selected?: T | T[];
+  sort?: SortProps;
+}
+
+function forceArray<T>(potentialArray: T | T[]): T[] {
+  return Array.isArray(potentialArray) ? potentialArray : [potentialArray];
+}
+
+const defaultProps = {
+  expandable: false,
+  expanded: [],
+  onPageChange: noop,
+  onSort: noop,
+  onVisibleRowsChange: noop,
+  page: 0,
+  pageSize: 10,
+  paginated: false,
+  selectable: false,
+  selected: [],
+  sort: { ascending: true },
+};
+
+function UncontrolledTable<T>({
+  allSelected,
+  columns,
+  expandable = defaultProps.expandable,
+  expanded = defaultProps.expanded,
+  onExpand,
+  onPageChange = defaultProps.onPageChange,
+  onSelect,
+  onSelectAll,
+  onSort = defaultProps.onSort,
+  onVisibleRowsChange = defaultProps.onVisibleRowsChange,
+  page = defaultProps.page,
+  pageSize = defaultProps.pageSize,
+  paginated = defaultProps.paginated,
+  rowExpanded,
+  rows,
+  selectable = defaultProps.selectable,
+  selected = defaultProps.selected,
+  sort = defaultProps.sort,
+  ...remainingProps
+}: UncontrolledTableProps<T>) {
+  const [sortInternal, setSortInternal] = useState(sort);
+
+  const [expandedInternal, setExpandedInternal] = useState(
+    forceArray(expanded)
+  );
+  // When the `expanded` prop changes, make sure the component state stays in sync
+  const [prevExpanded, setPrevExpanded] = useState(expanded);
+  if (prevExpanded !== expanded) {
+    setExpandedInternal(forceArray(expanded));
+    setPrevExpanded(expanded);
+  }
+
+  const [allSelectedInternal, setAllSelectedInternal] = useState(allSelected);
+  const [selectedInternal, setSelectedInternal] = useState(
+    forceArray(selected)
+  );
+  // Wnen the `selected` prop changes, make sure the component state stays in sync
+  const [prevSelected, setPrevSelected] = useState(selected);
```

Comment:
For these `prev` variables I think you can import `usePrevious` from `react-use` which is implemented using `useRef` under the hood, it'll cut down the number of renders in half